### PR TITLE
feat: enable RegExp test for revertedWith matcher

### PR DESCRIPTION
**Pull request by glinda93**
This MR will enable RegExp test for `revertedWith` matchers

For example, here is updated documentation block:

```
You can also test if revert message matches to a regular expression:

.. code-block:: ts

  await expect(token.checkRole('ADMIN'))
    .to.be.revertedWith(/AccessControl: account .* is missing role .*/);
```

### CODE REVIEW

**Review by rzadp — APPROVED** [review]
@glinda93 This looks awesome!

**Comment by rzadp** [comment]
@glinda93 BTW, we have a [discord](https://discord.gg/cSSmtdq7jr) server now for supporting community of Waffle and useDapp if you are interested.

File: waffle-chai/src/matchers/revertedWith.ts [hunk 1]
```diff
@@ -22,11 +22,14 @@ export function supportRevertedWith(Assertion: Chai.AssertionStatic) {
...
+        const isReverted = revertReason instanceof RegExp
+          ? revertReason.test(revertString)
+          : revertString === revertReason;
         this.assert(
-          revertString === revertReason,
+          isReverted,
           `Expected transaction to be reverted with "${revertReason}", but other reason was found: "${revertString}"`,
           `Expected transaction NOT to be reverted with "${revertReason}"`,
-          `Transaction reverted with "${revertReason}".`,
+          `Transaction reverted with "${revertReason}."`,
```
Comment (rzadp) [inline waffle-chai/src/matchers/revertedWith.ts:32]:
```suggestion
          `Transaction reverted with "${revertReason}"`,
```

**Comment by the-vampiire** [comment]
is this feature functional? i am having 2 issues:

1. the types are still reflecting only a string param
https://github.com/TrueFiEng/Waffle/blob/39c9eb04027fb5d5069db032493e378be3f3b2d0/waffle-chai/src/types.ts#L10
2. i am not able to match using a regex (after overriding the type to be `string | RegEx`

here is a contrived example that came up in class today:
```ts
// erc20 is OZ erc20 contract (no modifications)
await expect(
  erc20.decreaseAllowance(spenderAccount, -1000)
).to.be.revertedWith(/INVALID_ARGUMENT/);
```

> `AssertionError: Expected transaction to be reverted with /INVALID_ARGUMENT/, but other exception was thrown: Error: value out-of-bounds (argument="subtractedValue", value=-1000, code=INVALID_ARGUMENT, version=abi/5.6.0)`

**Comment by glinda93** [comment]
@the-vampiire Are you using the correct version? It's released in 3.3.0

**Comment by the-vampiire** [comment]
I think so. I am using hardhat and have in dev deps:

ethereum-waffle: ^3.4.4

Checking node modules I have:

@ethereum-waffle/chai version 3.4.4

**Comment by glinda93** [comment]
According to the code https://github.com/TrueFiEng/Waffle/blob/master/waffle-chai/src/matchers/revertedWith.ts#L37
you should have this message: 
> Expected transaction to be reverted with /INVALID_ARGUMENT/, but other reason was found:

**Comment by glinda93** [comment]
> I think so. I am using hardhat and have in dev deps:
> 
> ethereum-waffle: ^3.4.4
> 
> Checking node modules I have:
> 
> @ethereum-waffle/chai version 3.4.4

According to the code https://github.com/TrueFiEng/Waffle/blob/master/waffle-chai/src/matchers/revertedWith.ts#L37
you should have this message: 
> Expected transaction to be reverted with /INVALID_ARGUMENT/, but other reason was found:

**Comment by the-vampiire** [comment]
Strange man. That is not what I have here locally. Even uninstalled and installed again to be sure. 

Does master reflect the release? Maybe it is ahead of the 3.4.4 release?

**Comment by glinda93** [comment]
> Strange man. That is not what I have here locally. Even uninstalled and installed again to be sure.
> 
> Does master reflect the release? Maybe it is ahead of the 3.4.4 release?

@the-vampiire 3.4.4 was released 27 days ago, while this PR was merged 22 days ago. So `master` is ahead of release.

**Comment by the-vampiire** [comment]
Are there multiple waffles? Looking more closely I see that these deps I have are pointing at EthWorks/Waffle

How bizarre. It's practically the same codebase. Wouldn't have noticed had it not been for this issue

**Comment by glinda93** [comment]
> Are there multiple waffles? Looking more closely I see that these deps I have are pointing at EthWorks/Waffle
> 
> How bizarre. It's practically the same codebase. Wouldn't have noticed had it not been for this issue

`EthWorks/Waffle` is redirected to here. It's like the repository has been transferred

**Comment by the-vampiire** [comment]
these are my dev deps:

```json
"devDependencies": {
    "@nomiclabs/hardhat-ethers": "^2.0.5",
    "@nomiclabs/hardhat-etherscan": "^3.0.3",
    "@nomiclabs/hardhat-waffle": "^2.0.3",
    "@typechain/ethers-v5": "^7.2.0",
    "@typechain/hardhat": "^2.3.1",
    "@types/chai": "^4.3.0",
    "@types/mocha": "^9.1.0",
    "@types/node": "^12.20.47",
    "@typescript-eslint/eslint-plugin": "^4.33.0",
    "@typescript-eslint/parser": "^4.33.0",
    "chai": "^4.3.6",
    "dotenv": "^10.0.0",
    "eslint": "^7.32.0",
    "eslint-config-prettier": "^8.5.0",
    "eslint-config-standard": "^16.0.3",
    "eslint-plugin-import": "^2.25.4",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^3.4.1",
    "eslint-plugin-promise": "^5.2.0",
    "ethereum-waffle": "^3.4.4",
    "ethers": "^5.6.2",
    "hardhat": "^2.9.2",
    "hardhat-gas-reporter": "^1.0.8",
    "prettier": "^2.6.1",
    "prettier-plugin-solidity": "^1.0.0-beta.13",
    "solhint": "^3.3.7",
    "solidity-coverage": "^0.7.20",
    "ts-node": "^10.7.0",
    "typechain": "^5.2.0",
    "typescript": "^4.6.3"
  },
```

the steps i took were

1. `npm i -D hardhat`
2. `npx hardhat` > select "advanced typescript project"
3. run the install command directed by the CLI

```sh
npm install --save-dev "hardhat@^2.9.3" "@nomiclabs/hardhat-waffle@^2.0.0" "ethereum-waffle@^3.0.0" "chai@^4.2.0" "@nomiclabs/hardhat-ethers@^2.0.0" "ethers@^5.0.0" "@nomiclabs/hardhat-etherscan@^3.0.0" "dotenv@^10.0.0" "eslint@^7.29.0" "eslint-config-prettier@^8.3.0" "eslint-config-standard@^16.0.3" "eslint-plugin-import@^2.23.4" "eslint-plugin-node@^11.1.0" "eslint-plugin-prettier@^3.4.0" "eslint-plugin-promise@^5.1.0" "hardhat-gas-reporter@^1.0.4" "prettier@^2.3.2" "prettier-plugin-solidity@^1.0.0-beta.13" "solhint@^3.3.6" "solidity-coverage@^0.7.16" "@typechain/ethers-v5@^7.0.1" "@typechain/hardhat@^2.3.0" "@typescript-eslint/eslint-plugin@^4.29.1" "@typescript-eslint/parser@^4.29.1" "@types/chai@^4.2.21" "@types/node@^12.0.0" "@types/mocha@^9.0.0" "ts-node@^10.1.0" "typechain@^5.1.2" "typescript@^4.5.2"
```

**Comment by the-vampiire** [comment]
> `@ethereum-waffle/chai/package.json`

```json
{
  "name": "@ethereum-waffle/chai",
  "description": "A sweet set of chai matchers for your blockchain testing needs.",
  "version": "3.4.4",
  "author": "Marek Kirejczyk <account@ethworks.io> (http://ethworks.io)",
  "repository": "git@github.com:EthWorks/Waffle.git",
  "private": false,
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "ethereum",
    "smart-contracts",
    "solidity",
    "testing",
    "javascript",
    "typescript",
    "library"
  ],
  "homepage": "https://github.com/EthWorks/Waffle",
  "bugs": {
    "url": "https://github.com/EthWorks/Waffle/issues"
  },
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.ts",
  "types": "dist/esm/index.d.ts",
  "scripts": {
    "prepublishOnly": "yarn build",
    "test": "mocha",
    "lint": "eslint '{src,test}/**/*.ts'",
    "lint:fix": "eslint --fix '{src,test}/**/*.ts'",
    "build": "rimraf ./dist && yarn build:esm && yarn build:cjs",
    "build:esm": "tsc -p tsconfig.build.json --outDir dist/esm --module ES6",
    "build:cjs": "tsc -p tsconfig.build.json --outDir dist/cjs --declaration false"
  },
  "engines": {
    "node": ">=10.0"
  },
  "dependencies": {
    "@ethereum-waffle/provider": "^3.4.4",
    "ethers": "^5.5.2"
  },
  "devDependencies": {}
}
```

> `@ethereum-waffle/chai/dist/esm/types.d.ts`

```js
/// <reference types="chai" />
/// <reference types="sinon-chai" />
declare namespace Chai {
    interface Assertion extends LanguageChains, NumericComparison, TypeComparison {
        reverted: AsyncAssertion;
        revertedWith(reason: string): AsyncAssertion;
        emit(contract: any, eventName: string): EmitAssertion;
        properHex(length: number): void;
        hexEqual(other: string): void;
        properPrivateKey: void;
        properAddress: void;
        /**
         * @deprecated Use `changeEtherBalance()` instead.
         */
        changeBalance(account: any, balance: any): AsyncAssertion;
        /**
         * @deprecated Use `changeEtherBalances()` instead.
         */
        changeBalances(accounts: any[], balances: any[]): AsyncAssertion;
        changeEtherBalance(account: any, balance: any, options?: any): AsyncAssertion;
        changeEtherBalances(accounts: any[], balances: any[], options?: any): AsyncAssertion;
        changeTokenBalance(token: any, account: any, balance: any): AsyncAssertion;
        changeTokenBalances(token: any, accounts: any[], balances: any[]): AsyncAssertion;
        calledOnContract(contract: any): void;
        calledOnContractWith(contract: any, parameters: any[]): void;
    }
    interface NumberComparer {
        (value: any, message?: string): Assertion;
    }
    interface CloseTo {
        (expected: any, delta: any, message?: string): Assertion;
    }
    interface AsyncAssertion extends Assertion, Promise<void> {
    }
    interface EmitAssertion extends AsyncAssertion {
        withArgs(...args: any[]): AsyncAssertion;
    }
}
```

> `@ethereum-waffle/chai/dist/esm/matchers/revertedWith.js`

```js
export function supportRevertedWith(Assertion) {
    Assertion.addMethod('revertedWith', function (revertReason) {
        const promise = this._obj;
        const onSuccess = (value) => {
            this.assert(false, 'Expected transaction to be reverted', 'Expected transaction NOT to be reverted', 'Transaction reverted.', 'Transaction NOT reverted.');
            return value;
        };
        const onError = (error) => {
            // See https://github.com/ethers-io/ethers.js/issues/829
            const isEstimateGasError = error instanceof Object &&
                error.code === 'UNPREDICTABLE_GAS_LIMIT' &&
                'error' in error;
            if (isEstimateGasError) {
                error = error.error;
            }
            const reasonsList = error.results && Object.values(error.results).map((o) => o.reason);
            const message = (error instanceof Object && 'message' in error) ? error.message : JSON.stringify(error);
            const isReverted = reasonsList
                ? reasonsList.some((r) => r === revertReason)
                : message.includes('revert') && message.includes(revertReason);
            const isThrown = message.search('invalid opcode') >= 0 && revertReason === '';
            this.assert(isReverted || isThrown, `Expected transaction to be reverted with ${revertReason}, but other exception was thrown: ${error}`, `Expected transaction NOT to be reverted with ${revertReason}`, `Transaction reverted with ${revertReason}.`, error);
            return error;
        };
        const derivedPromise = promise.then(onSuccess, onError);
        this.then = derivedPromise.then.bind(derivedPromise);
        this.catch = derivedPromise.catch.bind(derivedPromise);
        return this;
    });
}
```

**Comment by the-vampiire** [comment]
sorry @glinda93 i missed this message 

> 3.4.4 was released 27 days ago, while this PR was merged 22 days ago. So master is ahead of release.

so i think this must be the issue. i was confused because i was reading the docs that said you could use regex. 

but you are saying that only the next release (once master is out) will include this feature?

**Comment by rzadp** [comment]
@the-vampiire You could use Waffle version alpha (currently `4.0.0-alpha.21`) that includes this feature, but beware because things might not be 100% stable.

**Pull request merged**